pkgs/http_client_conformance_tests/lib/src/request_methods_tests.dart

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:async/async.dart';
+import 'package:http/http.dart';
+import 'package:stream_channel/stream_channel.dart';
+import 'package:test/test.dart';
+
+import 'request_methods_server_vm.dart'
+    if (dart.library.html) 'request_methods_server_web.dart';
+
+/// Tests that the [Client] correctly sends HTTP request methods
+/// (e.g. GET, HEAD).
+void testRequestMethods(Client client) async {
+  group('request methods', () {
+    late final String host;
+    late final StreamChannel<Object?> httpServerChannel;
+    late final StreamQueue<Object?> httpServerQueue;
+
+    setUpAll(() async {
+      httpServerChannel = await startServer();
+      httpServerQueue = StreamQueue(httpServerChannel.stream);
+      host = 'localhost:${await httpServerQueue.next}';
+    });
+    tearDownAll(() => httpServerChannel.sink.add(null));
+
+    test('custom method', () async {
+      await client.send(Request(
+        'CUSTOM',
```

Comment:
Should we test case-sensitive method too?

Reply:
The problem is that IOClient would not pass that test because it always uppercases the request method:
https://github.com/dart-lang/sdk/blob/0b10bfcde99def6c43d4a306c920ba1ba855d058/sdk/lib/_http/http_impl.dart#L2791

This seems like a bug according to RFC-7232: https://www.rfc-editor.org/rfc/rfc7231#section-4.1

I will add a case-insensitive test and then a flag to disable it.